File: go/execution-kernel/internal/router/advisor.go
package router

import (
	"context"
	"encoding/json"
	"fmt"
	"os/exec"
	"strings"
	"time"
)

// Why an advisor here at all, given that Hermes ships a "smart mode"
// approval gate (`approvals.mode: smart` in Hermes' `config.yaml`)?
// The two systems answer different questions and live at different
// layers of the stack. Hermes' smart mode is a per-tool-call,
// per-Hermes-driver gate that decides whether to surface an approval
// prompt to the operator on dangerous patterns; it is a routing
// decision local to one driver's loop and authoritative only inside
// Hermes' own runtime.
//
// chitin's advisor is upstream of any specific driver. It sits behind
// `chitin-kernel router evaluate --hook-stdin`, which is wired into
// every driver chitin gates (Claude Code, Codex, Gemini, Hermes
// itself). It runs only when a kernel-level heuristic (blast radius,
// floundering chain, plugin signal) crosses a threshold the policy
// declared as advisor-worthy, and its output is a structured
// continue/takeover verdict plus a nudge — not an operator prompt.
// The verdict feeds the chain envelope and gov-decisions audit log
// the same way any other gate decision does.
//
// The asymmetry is deliberate. Hermes' smart mode is the right home
// for "ask the human now" because Hermes owns the chat surface where
// the human already is. chitin's advisor is the right home for
// "ask another model whether this looks off" because chitin owns the
// cross-driver canonical-action vocabulary the heuristics score
// against. Neither subsumes the other; collapsing them would either
// force chitin to re-implement Hermes' gateway (the mistake culled
// on 2026-05-08, see
// docs/decisions/2026-05-08-cull-escalate-defer-to-hermes.md) or
// force Hermes' approval gate to understand chitin's chain-replay
// semantics. The advisor justifies itself by being the only LLM
// second-opinion that runs across all drivers from one canonical
// action vocabulary; smart mode justifies itself by owning the chat.

// CallAdvisor spawns `claude -p` (sub-billed Claude Code Pro plan)
// with a structured advisor prompt and parses the
// <<<ROUTER_ADVISOR>>>{...} marker from stdout. Returns
// (response, nil) on success, (nil, err) on failure.
//
// Constraint: NO metered API calls. We use the Claude Code CLI's
// `claude -p` interface which authenticates against the user's
// Claude Code subscription. Per the Anthropic AUP clarification
// (2026-05-02), headless `claude -p` is supported for automation.
func CallAdvisor(ctx context.Context, req AdvisorRequest, binary string, timeout time.Duration) (*AdvisorResponse, error) {
	if binary == "" {
		binary = "claude"
	}
	if timeout == 0 {
		timeout = 60 * time.Second
	}
	prompt := buildAdvisorPrompt(req)
	cctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	cmd := exec.CommandContext(cctx, binary, "-p", "--dangerously-skip-permissions", "--output-format", "text")
	cmd.Stdin = strings.NewReader(prompt)
	out, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("advisor spawn failed: %w", err)
	}
	resp := parseAdvisorOutput(string(out))
	if resp == nil {
		return nil, fmt.Errorf("advisor output missing or malformed marker")
	}
	return resp, nil
}

func buildAdvisorPrompt(req AdvisorRequest) string {
	heur, _ := json.MarshalIndent(req.HeuristicOutcome, "", "  ")
	action, _ := json.MarshalIndent(req.ProposedAction, "", "  ")
	return fmt.Sprintf(`You are a chitin ROUTER ADVISOR — called when a lower-tier agent's tool call triggered a heuristic threshold.

Your job: decide whether the agent should proceed (with what nudge), OR whether the action should be blocked / escalated to a higher-tier human takeover.

PROPOSED ACTION:
%s

HEURISTIC OUTCOME:
%s

CALLER CONTEXT:
- caller_tier: %s
- chain_depth: %d (max chain_depth before takeover: typically 3)

QUESTION:
%s

CONTEXT:
%s

YOUR OUTPUT — emit EXACTLY this JSON line as your final output (no other text after it):

<<<ROUTER_ADVISOR>>>{"nudge": "<short text shown to the agent>", "verdict": "continue"|"takeover", "escalate": <bool>}
`,
		string(action), string(heur), req.CallerTier, req.ChainDepth,
		req.Question, req.Context,
	)
}

func parseAdvisorOutput(stdout string) *AdvisorResponse {
	const marker = "<<<ROUTER_ADVISOR>>>"
	lastIdx := strings.LastIndex(stdout, marker)
	if lastIdx < 0 {
		return nil
	}
	rest := stdout[lastIdx+len(marker):]
	if !strings.HasPrefix(rest, "{") {
		return nil
	}
	depth := 0
	end := -1
	for i, ch := range rest {
		if ch == '{' {
			depth++
		} else if ch == '}' {
			depth--
			if depth == 0 {
				end = i + 1
				break
			}
		}
	}
	if end < 0 {
		return nil
	}
	var resp AdvisorResponse
	if err := json.Unmarshal([]byte(rest[:end]), &resp); err != nil {
		return nil
	}
	if resp.Verdict != "continue" && resp.Verdict != "takeover" {
		return nil
	}
	return &resp
}

File: apps/runner/src/router/advisor.ts
// Advisor primitive — wraps `claude -p` (subscription-billed
// headless mode) to produce structured advice for the router's
// hook wrapper.
//
// Constraint: NO metered API calls. We use the Claude Code CLI's
// `claude -p "prompt"` interface which authenticates against the
// user's Claude Code subscription (Pro plan). Per the Anthropic
// AUP clarification (2026-05-02), headless `claude -p` is
// supported for automation.
//
// MVP: single advisor call per invocation. Chain depth is managed
// by the hook wrapper, which can call this multiple times with
// incrementing chain_depth.

import { spawn } from 'node:child_process';
import type {
  AdvisorRequest,
  AdvisorResponse,
  HeuristicOutcome,
} from './types.ts';

/** Build the advisor prompt from a structured request. */
export function buildAdvisorPrompt(request: AdvisorRequest): string {
  const heuristic_summary = renderHeuristicSummary(request.heuristic_outcome);
  return `You are a chitin ROUTER ADVISOR — called when a lower-tier agent's tool call triggered a heuristic threshold.

Your job: decide whether the agent should proceed, with what nudge, OR whether the action should be blocked / escalated to a higher-tier human takeover.

PROPOSED ACTION:
\`\`\`json
${JSON.stringify(request.proposed_action, null, 2)}
\`\`\`

HEURISTIC OUTCOME (which signals fired):
${heuristic_summary}

CALLER CONTEXT:
- caller_tier: ${request.caller_tier ?? 'unknown'}
- chain_depth: ${request.chain_depth} (max chain_depth before takeover: typically 3)

QUESTION:
${request.question}

CONTEXT:
${request.context}

YOUR OUTPUT — emit EXACTLY this JSON line as your final output (no other text after it):

\`\`\`
<<<ROUTER_ADVISOR>>>{"nudge": "<short text shown to the agent>", "verdict": "continue"|"takeover", "escalate": <bool>}
\`\`\`

Where:
- nudge: short specific text (1-3 sentences) the agent will see in its tool result. Focus on the SPECIFIC concern that fired the heuristic. Don't lecture; nudge.
- verdict: "continue" lets the action proceed (with nudge attached); "takeover" blocks it and signals re-dispatch at higher tier.
- escalate: true if you think a higher-tier advisor should be consulted INSTEAD of you (chain to T+1). Use sparingly — only for cases where you don't have the domain knowledge.

Examples of good nudges:
- "Consider running tests before this commit — the diff touches 12 files."
- "This is a force-push on a branch with open PRs (#142, #143). Verify intent."
- "You've tried this same edit 3 times with the same syntax error. The actual issue is X."

Examples of bad nudges:
- "Be careful." (too vague)
- "Don't do this." (no reasoning)
- A multi-paragraph explanation (too verbose for an inline tool message)
`;
}

function renderHeuristicSummary(outcome: HeuristicOutcome): string {
  const parts: string[] = [];
  if (outcome.blast_radius) {
    parts.push(
      `- blast_radius: ${outcome.blast_radius.score} (${outcome.blast_radius.reason})${outcome.blast_radius.fired ? ' [FIRED]' : ''}`,
    );
  }
  if (outcome.drift) {
    parts.push(
      `- drift: ${outcome.drift.score} (${outcome.drift.reason})${outcome.drift.fired ? ' [FIRED]' : ''}`,
    );
  }
  if (outcome.floundering) {
    parts.push(
      `- floundering: ${outcome.floundering.score} (${outcome.floundering.reason})${outcome.floundering.fired ? ' [FIRED]' : ''}`,
    );
  }
  return parts.length > 0 ? parts.join('\n') : '(no heuristic ran — caller invoked advisor directly)';
}

/** Pure: parse the advisor's structured output marker. */
export function parseAdvisorOutput(stdout: string): AdvisorResponse | null {
  const marker = /<<<ROUTER_ADVISOR>>>(\{[\s\S]*?\})/;
  const matches = [...stdout.matchAll(new RegExp(marker, 'g'))];
  if (matches.length === 0) return null;
  const last = matches[matches.length - 1];
  try {
    const parsed = JSON.parse(last[1]) as Record<string, unknown>;
    if (typeof parsed.nudge !== 'string' || typeof parsed.verdict !== 'string') {
      return null;
    }
    if (parsed.verdict !== 'continue' && parsed.verdict !== 'takeover') {
      return null;
    }
    return {
      nudge: parsed.nudge,
      verdict: parsed.verdict,
      escalate: typeof parsed.escalate === 'boolean' ? parsed.escalate : false,
    };
  } catch {
    return null;
  }
}

/**
 * Run the advisor. I/O: spawns `claude -p` with the prompt,
 * waits for completion (bounded by timeout), parses the output
 * marker. Returns null on any failure (timeout, parse error,
 * binary missing) — caller decides fallback (typically: pass-
 * through to deterministic kernel verdict).
 */
export async function callAdvisor(
  request: AdvisorRequest,
  opts: { timeoutMs?: number; binary?: string } = {},
): Promise<AdvisorResponse | null> {
  const timeoutMs = opts.timeoutMs ?? 60_000;
  const binary = opts.binary ?? 'claude';
  const prompt = buildAdvisorPrompt(request);

  return new Promise((resolve) => {
    let resolved = false;
    const finish = (val: AdvisorResponse | null): void => {
      if (resolved) return;
      resolved = true;
      resolve(val);
    };

    const child = spawn(
      binary,
      ['-p', '--dangerously-skip-permissions', '--output-format', 'text'],
      { stdio: ['pipe', 'pipe', 'pipe'] },
    );
    let stdout = '';
    let stderr = '';
    child.stdout.on('data', (d: Buffer) => (stdout += d.toString('utf8')));
    child.stderr.on('data', (d: Buffer) => (stderr += d.toString('utf8')));

    const timer = setTimeout(() => {
      child.kill('SIGKILL');
      // Log the timeout to stderr; caller's logging happens upstream
      console.error(
        JSON.stringify({
          ts: new Date().toISOString(),
          level: 'warn',
          component: 'router-advisor',
          msg: 'advisor-timeout',
          timeoutMs,
        }),
      );
      finish(null);
    }, timeoutMs);

    child.on('error', () => {
      clearTimeout(timer);
      console.error(
        JSON.stringify({
          ts: new Date().toISOString(),
          level: 'warn',
          component: 'router-advisor',
          msg: 'advisor-spawn-error',
          stderr: stderr.slice(0, 500),
        }),
      );
      finish(null);
    });
    child.on('close', () => {
      clearTimeout(timer);
      finish(parseAdvisorOutput(stdout));
    });

    // Send the prompt over stdin
    child.stdin.write(prompt);
    child.stdin.end();
  });
}
